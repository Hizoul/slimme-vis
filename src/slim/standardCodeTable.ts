import isEqual from "../util/isEqual"
import { Database } from "./processDatabaseAsString"

export interface ICodeTable {
  left: any
  right: any
}

export interface CodeTable {[index: string]: number}
export type ICodeTableEntry = number[][]
export type SortedCodeTable = ICodeTableEntry[]

const standardCodeTable = (dataset: Database) => {
  const itemSets: CodeTable = {}
  for (const key of Object.keys(dataset)) {
    for (const itemSet of dataset[key]) {
      for (const value of itemSet) {
        const itemKey = String(value)
        if (itemSets[itemKey] == null) {
          itemSets[itemKey] = 0
        }
        itemSets[itemKey]++
      }
    }
  }
  return itemSets
}

const updateFrequencies = (dataset: Database, codeTable: SortedCodeTable) => {
  const clonedDb: Database = JSON.parse(JSON.stringify(dataset))
  for (const entry of codeTable) {
    entry[1][0] = 0
    for (const classKey of Object.keys(clonedDb)) {
      for (const transactionIndex in clonedDb[classKey]) {
        const transaction = clonedDb[classKey][transactionIndex]
        if (subset(entry[0], transaction).length === entry[0].length) {
          entry[1][0]++
          clonedDb[classKey][transactionIndex] = transaction.filter((value) => entry[0].indexOf(value) === -1)
        }
      }
    }
  }
  return standardCoverOrder(codeTable)
}

const supportOfItem = (itemSet: number[], dataset: Database) => {
  let support = 0
  for (const key of Object.keys(dataset)) {
    for (const set of dataset[key]) {
      if (subset(itemSet, set).length === itemSet.length) {
        support++
      }
    }
  }
  return support
}

const sortedStandardCodeTable = (dataset: Database) => {
  const itemSets: CodeTable = standardCodeTable(dataset)
  const sortedCodeTable: SortedCodeTable = []
  for (const key of Object.keys(itemSets)) {
    sortedCodeTable.push([[Number(key)], [itemSets[key]], [supportOfItem([itemSets[key]], dataset)]])
  }
  return standardCoverOrder(sortedCodeTable)
}

// first descending on cardinality (amount of values), second descending on support (here called frequency), and third lexicographically
const standardCoverOrder = (codeTable: SortedCodeTable) => {
  codeTable.sort((itemA, itemB) => {
    if (itemA[0].length < itemB[0].length) {
      return 1
    }
    if (itemA[0].length > itemB[0].length) {
      return -1
    }
    if (itemA[2][0] < itemB[2][0]) {
      return 1
    }
    if (itemA[2][0] > itemB[2][0]) {
      return -1
    }
    return -1 * JSON.stringify(itemA[0]).localeCompare(JSON.stringify(itemB[0]))
  })
  return codeTable
}

const subset = (arrA: any[], arrB: any[]) => {
  const arrToUse = arrA.length > arrB.length ? arrB : arrA
  const arrToCompare = arrA.length > arrB.length ? arrA : arrB
  const result: any[] = []
  arrToUse.map((value) => {
    if (arrToCompare.indexOf(value) !== -1) {
      result.push(value)
    }
  })
  return result
}

export default sortedStandardCodeTable
export {
  standardCoverOrder, subset, updateFrequencies, supportOfItem
}
