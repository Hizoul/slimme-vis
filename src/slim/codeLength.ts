import isEqual from "../util/isEqual"
import log2 from "../util/log"
import coverCode from "./coverCode"
import { Database } from "./processDatabaseAsString"
import { CodeTable, SortedCodeTable } from "./standardCodeTable"

const itemProbability = (codeTable: SortedCodeTable, set: number[]) => {
  let top = 0
  let fittingEntry: any
  let frequencySum = 0
  codeTable.forEach((entry) => {
    frequencySum += entry[1][0]
    if (isEqual(entry[0], set)) {
      fittingEntry = entry
    }
  })
  if (fittingEntry != null) {
    top = fittingEntry[1][0]
  }
  return top / frequencySum
}

const lengthForSet = (codeTable: SortedCodeTable, set: number[]) => {
  return -log2(itemProbability(codeTable, set))
}

const lengthForSetInStandard = (codeTable: SortedCodeTable, set: number[]) => {
  const cover = coverCode(codeTable, set)
  let coverSum = 0
  for (const code of cover) {
    coverSum += lengthForSet(codeTable, code)
  }
  return coverSum
}

const codeLength = (dataset: Database, codeTable: SortedCodeTable) => {
  let sum = 0
  for (const classKey of Object.keys(dataset)) {
    for (const itemSet of dataset[classKey]) {
      const cover = coverCode(codeTable, itemSet)
      let coverSum = 0
      for (const code of cover) {
        coverSum += lengthForSet(codeTable, code)
      }
      sum += coverSum
    }
  }
  return sum
}
function isNumber(n: number) {
  return typeof n === "number" && !isNaN(n) && isFinite(n)
}
const lengthWithStandard = (dataset: Database, codeTable: SortedCodeTable, standardTable: SortedCodeTable) => {
  let sum = 0
  for (const entry of codeTable) {
    const a = lengthForSet(codeTable, entry[0])
    const b = lengthForSet(standardTable, entry[0])
    if (isNumber(a)) {
      sum += a
    }
    if (isNumber(b)) {
      sum += b
    }
  }
  return sum
}

const codeLengthIncludingStandard = (dataset: Database, codeTable: SortedCodeTable, standardTable: SortedCodeTable) => {
  return lengthWithStandard(dataset, codeTable, standardTable) + codeLength(dataset, codeTable)
}

export default codeLength
export {
  codeLengthIncludingStandard, lengthForSet, lengthForSetInStandard
}
