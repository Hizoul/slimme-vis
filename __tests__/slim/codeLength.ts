import { sha512 } from "js-sha512"
import { resolve } from "path"
import processLocalDb from "../../src/node/local"
import codeLength, { codeLengthIncludingStandard } from "../../src/slim/codeLength"
import standardCodeTable, { ICodeTableEntry, standardCoverOrder } from "../../src/slim/standardCodeTable"
import sortedStandardCodeTable from "../../src/slim/standardCodeTable"

describe("ProcessDB", () => {
  it("code length", () => {
    const dataset = processLocalDb(resolve("__tests__/test.dat"))
    const sorted = sortedStandardCodeTable(dataset)
    const diff: ICodeTableEntry[] = [
      [[ 6, 7 ], [3]],
      [[ 5 ], [3 ]],
      [[ 8 ], [2 ]],
      [[ 3 ], [2 ]],
      [[ 13 ], [2 ]],
      [[ 12 ], [2 ]],
      [[ 10 ], [2 ]],
      [[ 4 ], [1 ]],
      [[ 6 ], [1]],
      [[ 7 ], [1]],
      [[ 14 ], [1 ]],
      [[ 11 ], [1 ]]
    ]
    expect(codeLength(dataset, sorted)).toMatchSnapshot("result of initialLength")
    expect(codeLength(dataset, diff)).toMatchSnapshot("result of diffLength")
    expect(codeLengthIncludingStandard(dataset, diff, sorted)).toMatchSnapshot("result of difftable")
  })
})
