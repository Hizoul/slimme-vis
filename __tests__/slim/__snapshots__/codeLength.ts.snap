// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessDB code length: result of diffLength 1`] = `72.72889087402703`;

exports[`ProcessDB code length: result of difftable 1`] = `157.1163999531414`;

exports[`ProcessDB code length: result of initialLength 1`] = `79.28421251514429`;
